neps/nep-0417.md

```diff
@@ -0,0 +1,202 @@
+---
+NEP: 417
+Title: Meta Transactions
+Author: Illia Polosukhin <ilblackdragon@gmail.com>, Egor Uleyskiy (egor.ulieiskii@gmail.com), Alexander Fadeev (fadeevab.com@gmail.com)
+DiscussionsTo: https://github.com/nearprotocol/neps/pull/417
+Status: Draft
+Type: Protocol Track
+Category: Runtime
+Created: 19-Oct-2022
+---
+
+## Summary
+
+In-protocol meta transactions allowing for third-party account to initiate and pay transaction fees on behalf of the account.
+
+## Motivation
+
+NEAR has been designed with simplicity of onboarding in mind. One of the large hurdles right now is that after creating an implicit or even named account the user does not have NEAR to pay gas fees to interact with apps.
+
+For example, apps that pay user for doing work (like NEARCrowd or Sweatcoin) or free-to-play games.
+
+[Aurora Plus](https://aurora.plus) has shown viability of the relayers that can offer some number of free transactions and a subscription model. Shifting the complexity of dealing with fees to the infrastructure from the user space.
+
+## Rationale and alternatives
+
+The proposed design here provides the easiest way for users and developers to onboard and to pay for user transactions.
+There are no requirements on the user account. The user account may not even exist on the chain and an implicit account can be used instead.
+
+An alternative is to a proxy contracts deployed on the user account.
+This design has severe limitations as it requires the user to deploy such contract and incur additional costs for storage.
+
+## Specification
+
+The main flow of the meta transaction will be as follows:
+ - User requests the relayer for its access key (public key) which the relayer are going to use. The user specifies this key as `relayer_pub_key` in `DelegateAction`.
+ - User fetches the relayer's access key `nonce`, increments it and assigns the value to `relayer_nonce` in `DelegateAction`: `relayer_nonce = nonce + 1`.
```

Comment:
Will the relayer have one nonce per user?  If not, then a relayer can only work with one user at a time.  This will have lower throughput and a potential attack vector where the user can request the nonce from the relayer and never send the subsequent transaction.

Reply:
Actually the implementation of the relayer node is up to developers, I think. But if you want me to describe it in this NEP, I will do. Bellow are my thoughts:

A relayer should have one nonce per access key. And one access key per user. A relayer can import a lot of access keys to its account and schedule them between users.

> the user can request the nonce from the relayer and never send the subsequent transaction

The relayer can reserve an access key for some time. When this time ends, other users can use this key.
And one important thing, the user requests *access key/public key*, not nonce.

Reply:
I agree that the implementation of the relayer node will be up to the developers.  And I suppose a NEP is not the right place to provide best practices for developers either.  However, I felt that fellow developers who are interested in building a relayer would be curious about this point.  For me personally, just this conversation should suffice and I don't think we need to add more text to the NEP.